pkg/kinds/general.go

```diff
@@ -152,17 +152,211 @@ func (m *GrafanaResourceMetadata) GetOriginInfo() *ResourceOriginInfo {
+	if err != nil {
+		return nil
+	}
+	return &t
+}
+
+// Accessor functions for k8s objects
+type GrafanaResourceMetaAccessor interface {
+	GetUpdatedTimestamp() *time.Time
+	SetUpdatedTimestamp(v *time.Time)
+	GetCreatedBy() string
+	SetCreatedBy(user string)
+	GetUpdatedBy() string
+	SetUpdatedBy(user string)
+	GetFolder() string
+	SetFolder(uid string)
+	GetSlug() string
+	SetSlug(v string)
+	GetOriginInfo() *ResourceOriginInfo
+	SetOriginInfo(info *ResourceOriginInfo)
+	GetOriginName() string
+	GetOriginPath() string
+	GetOriginKey() string
+	GetOriginTimestamp() *time.Time
+}
+
+var _ GrafanaResourceMetaAccessor = (*grafanaResourceMetaAccessor)(nil)
+var _ GrafanaResourceMetaAccessor = (*GrafanaResourceMetadata)(nil)
+
+type grafanaResourceMetaAccessor struct {
+	obj v1.Object
+}
+
+func MetaAccessor(obj v1.Object) GrafanaResourceMetaAccessor {
+	return &grafanaResourceMetaAccessor{obj}
+}
+
+func (m *grafanaResourceMetaAccessor) set(key string, val string) {
+	anno := m.obj.GetAnnotations()
+	if val == "" {
+		if anno != nil {
+			delete(anno, key)
 		}
+	} else {
+		if anno == nil {
+			anno = make(map[string]string)
+		}
+		anno[key] = val
+	}
+	m.obj.SetAnnotations(anno)
+}
+
+func (m *grafanaResourceMetaAccessor) get(key string) string {
+	return m.obj.GetAnnotations()[key]
+}
+
+func (m *grafanaResourceMetaAccessor) GetUpdatedTimestamp() *time.Time {
+	v, ok := m.obj.GetAnnotations()[annoKeyUpdatedTimestamp]
+	if !ok {
+		return nil
+	}
+	t, err := time.Parse(time.RFC3339, v)
+	if err != nil {
+		return nil
+	}
+	return &t
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedTimestampMillis(v int64) {
+	if v > 0 {
+		t := time.UnixMilli(v)
+		m.SetUpdatedTimestamp(&t)
+	} else {
+		m.SetUpdatedTimestamp(nil)
+	}
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedTimestamp(v *time.Time) {
+	txt := ""
+	if v != nil {
+		txt = v.UTC().Format(time.RFC3339)
+	}
+	m.set(annoKeyUpdatedTimestamp, txt)
+}
+
+func (m *grafanaResourceMetaAccessor) GetCreatedBy() string {
+	return m.get(annoKeyCreatedBy)
+}
+
+func (m *grafanaResourceMetaAccessor) SetCreatedBy(user string) {
+	m.set(annoKeyCreatedBy, user)
+}
+
+func (m *grafanaResourceMetaAccessor) GetUpdatedBy() string {
+	return m.get(annoKeyUpdatedBy)
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedBy(user string) {
+	m.set(annoKeyUpdatedBy, user)
+}
+
+func (m *grafanaResourceMetaAccessor) GetFolder() string {
+	return m.get(annoKeyFolder)
+}
+
+func (m *grafanaResourceMetaAccessor) SetFolder(uid string) {
+	m.set(annoKeyFolder, uid)
+}
+
+func (m *grafanaResourceMetaAccessor) GetSlug() string {
+	return m.get(annoKeySlug)
+}
+
+func (m *grafanaResourceMetaAccessor) SetSlug(v string) {
+	m.set(annoKeySlug, v)
+}
+
+func (m *grafanaResourceMetaAccessor) SetOriginInfo(info *ResourceOriginInfo) {
+	anno := m.obj.GetAnnotations()
+	if anno == nil {
+		if info == nil {
+			return
+		}
+		anno = make(map[string]string, 0)
+		m.obj.SetAnnotations(anno)
```

Comment:
technically this line isn't needed

Reply:
I did not think so, but without it the attached tests do not pass 🤷🏻 -- with unstructured.Unstructured the annotations you get from `GetAnnotations()` are a copy of the underlying data

Reply:
that is interesting, since we call m.obj.SetAnnotations at the end of the function.  Either way it doesn't do any harm.